python/paddle/fluid/tests/unittests/test_gradient_clip.py

```diff
@@ -84,78 +90,394 @@ def check_operators(self, place):
-        for data in train_reader():
-            count += 1
-            if count > 5:
-                break
-            out = exe.run(prog, feed=feeder.feed(data), fetch_list=grad_list)
-            out_clip = exe.run(prog_clip,
-                               feed=feeder.feed(data),
-                               fetch_list=grad_clip_list)
-            global_norm = 0
-            for v in out:
-                global_norm += np.sum(np.power(v, 2))
-            global_norm = np.sqrt(global_norm)
-
-            global_norm_clip = 0
-            for v in out_clip:
-                global_norm_clip += np.sum(np.power(v, 2))
-            global_norm_clip = np.sqrt(global_norm_clip)
-
-            assert np.isclose(
-                a=global_norm_clip, b=np.minimum(global_norm, CLIP), rtol=5e-3)
-
-    def check_sparse_gradient_clip(self, place):
-        prog = fluid.framework.Program()
-        startup_program = fluid.framework.Program()
-        with fluid.program_guard(
-                main_program=prog, startup_program=startup_program):
-            data = fluid.layers.data(
-                name="words", shape=[1], dtype="int64", lod_level=1)
-            label = fluid.layers.data(name="label", shape=[1], dtype="int64")
-            cost = bow_net(data, label, self.word_dict_len)
+        data = next(train_reader())
+        out = exe.run(prog, feed=feeder.feed(data), fetch_list=grad_list)
+        out_clip = exe.run(prog_clip,
+                           feed=feeder.feed(data),
+                           fetch_list=grad_clip_list)
+        self.check_output(out, out_clip)
 
-            fluid.clip.set_gradient_clip(
-                clip=fluid.clip.GradientClipByGlobalNorm(clip_norm=5.0))
+    def backward_and_optimize(cost):
+        pass
 
-            sgd_optimizer = fluid.optimizer.SGD(learning_rate=0.01)
-            sgd_optimizer.minimize(cost)
+    def check_with_optimize(self, place):
+        prog = fluid.Program()
+        startup_program = fluid.Program()
+        with fluid.program_guard(
+                main_program=prog, startup_program=startup_program):
+            words = fluid.data(
+                name="words", shape=[-1, 1], dtype="int64", lod_level=1)
+            label = fluid.data(name="label", shape=[-1, 1], dtype="int64")
+            cost = bow_net(words, label, self.word_dict_len)
+            self.backward_and_optimize(cost)
 
         exe = fluid.Executor(place)
-        feeder = fluid.DataFeeder(feed_list=[data, label], place=place)
+        feeder = fluid.DataFeeder(feed_list=[words, label], place=place)
         exe.run(startup_program)
 
         data = next(self.train_data())
         val = exe.run(prog, feed=feeder.feed(data), fetch_list=[cost])[0]
-        self.assertEqual((1, ), val.shape)
-        print(val)
-        self.assertFalse(np.isnan(val))
+        self.assertEqual((1, ), val.shape, "the shape of loss is not [1]")
+        self.assertFalse(np.isnan(val), "the loss of network is nan")
```

Comment:
Try to check more instead of only checking whether loss is nan? For example, check whether the gradient clip result is the expected one (which is from another calculation method) ?

Reply:
the gradient clip result has been checked by many other unittests